### PR TITLE
Add BoxCox and YeoJohnson primitives for time series problems

### LINKED ISSUE #2070: Add BoxCox primitive
- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.boxcox.html
scipy-stats-boxcox-1

### LINKED ISSUE #2071: Add Yeo-Johnson primitive
- https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.yeojohnson.html

### PR DESCRIPTION
**Pull request by gsheni**
- Closes #2070
- Closes #2071

### CODE REVIEW

File: featuretools/tests/primitive_tests/test_transform_primitive.py [hunk 1]
```diff
@@ -950,3 +956,28 @@ def test_rate_of_change_primitive_with_nan():
...
+
+@pytest.mark.parametrize(
+    "lmbda,expected_output",
+    [
+        (None, [0, 1, -5, 427, -114, -475]),
+        (-1, [0, 0, -443, 1, -334334333, -333433343333]),
+    ],
+)
+def test_yeo_johnson(lmbda, expected_output):
+    box_cox = YeoJohnson(lmbda=lmbda)
```
Comment (sbadithe) [inline featuretools/tests/primitive_tests/test_transform_primitive.py:980]:
```suggestion
    yeo_johnson = YeoJohnson(lmbda=lmbda)
```

File: featuretools/tests/primitive_tests/test_transform_primitive.py [hunk 2]
```diff
@@ -950,3 +956,28 @@ def test_rate_of_change_primitive_with_nan():
...
+    numeric_values = pd.Series([0, 1, -10, 100, -1000, -10000])
+    actual = box_cox(numeric_values)
```
Comment (sbadithe) [inline featuretools/tests/primitive_tests/test_transform_primitive.py:982]:
```suggestion
    actual = yeo_johnson(numeric_values)
```

File: featuretools/primitives/standard/transform/time_series/box_cox.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        >>> transformed
+        [0.0, 0.9, 0.8, 0.75, 0.5]
+
+    """
+
+    name = "box_cox"
+    input_types = [
+        ColumnSchema(semantic_tags={"numeric"}),
+    ]
+    return_type = ColumnSchema(semantic_tags={"numeric"})
```
Comment (sbadithe) [inline featuretools/primitives/standard/transform/time_series/box_cox.py:39]: I think this can be Double logical type

File: featuretools/primitives/standard/transform/time_series/yeo_johnson.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        >>> transformed
+        [0.5, -443.33, 0.83, -41.33, 0.67]
+
+    """
+
+    name = "yeo_johnson"
+    input_types = [
+        ColumnSchema(semantic_tags={"numeric"}),
+    ]
+    return_type = ColumnSchema(semantic_tags={"numeric"})
```
Comment (sbadithe) [inline featuretools/primitives/standard/transform/time_series/yeo_johnson.py:38]: Same here, I think the return type can be Double logical type

File: featuretools/tests/primitive_tests/test_transform_features.py [hunk 5]
```diff
@@ -121,7 +122,7 @@ def test_init_and_name(es):
     for transform_prim in trans_primitives:
         # skip automated testing if a few special cases
         features_to_use = log_features
-        if transform_prim in [NotEqual, Equal]:
+        if transform_prim in [NotEqual, Equal, BoxCox]:
```
Comment (sbadithe) [inline featuretools/tests/primitive_tests/test_transform_features.py:125]: curious, why is BoxCox here but not Yeo-Johnson?

Comment (gsheni) [reply featuretools/tests/primitive_tests/test_transform_features.py:126]: Put comment here why:
https://github.com/alteryx/featuretools/pull/2387/commits/475efc8709837593cbea0a7194cb90453f25dd6b

File: featuretools/primitives/standard/transform/time_series/yeo_johnson.py (hunk 4 above)
Comment (gsheni) [reply featuretools/primitives/standard/transform/time_series/yeo_johnson.py]: Fixed
https://github.com/alteryx/featuretools/pull/2387/commits/475efc8709837593cbea0a7194cb90453f25dd6b

File: featuretools/primitives/standard/transform/time_series/box_cox.py (hunk 3 above)
Comment (gsheni) [reply featuretools/primitives/standard/transform/time_series/box_cox.py]: Fixed
https://github.com/alteryx/featuretools/pull/2387/commits/475efc8709837593cbea0a7194cb90453f25dd6b

File: featuretools/tests/primitive_tests/test_transform_primitive.py (hunk 2 above)
Comment (gsheni) [reply featuretools/tests/primitive_tests/test_transform_primitive.py]: Fixed
https://github.com/alteryx/featuretools/pull/2387/commits/475efc8709837593cbea0a7194cb90453f25dd6b

File: featuretools/tests/primitive_tests/test_transform_primitive.py (hunk 1 above)
Comment (gsheni) [reply featuretools/tests/primitive_tests/test_transform_primitive.py]: Fixed:
https://github.com/alteryx/featuretools/pull/2387/commits/475efc8709837593cbea0a7194cb90453f25dd6b

File: featuretools/tests/primitive_tests/test_transform_features.py (hunk 5 above)
Comment (gsheni) [reply featuretools/tests/primitive_tests/test_transform_features.py:126]: added comment why

File: featuretools/primitives/standard/transform/email/is_free_email_domain.py [hunk 6]
```diff
@@ -36,7 +36,7 @@ def get_function(self):
         def is_free_email_domain(emails):
             # if the input is empty return an empty Series
             if len(emails) == 0:
-                return pd.Series([])
+                return pd.Series([], dtype=BooleanNullable.primary_dtype)
```
Comment (gsheni) [inline featuretools/primitives/standard/transform/email/is_free_email_domain.py:39]:
```suggestion
                return pd.Series([], dtype="boolean")
```

File: docs/source/release_notes.rst [hunk 7]
```diff
@@ -8,14 +8,15 @@ Future Release
     * Enhancements
         * Add ``OneDigitPostalCode`` and ``TwoDigitPostalCode`` primitives (:pr:`2365`)
         * Add ``ExpandingCount``, ``ExpandingMin``, ``ExpandingMean``, ``ExpandingMax``, ``ExpandingSTD``, and ``ExpandingTrend`` primitives (:pr:`2343`)
+        * Add ``BoxCox`` and ``YeoJohnson`` primitive for time series problems (:pr:`2387`)
```
Comment (thehomebrewnerd) [inline docs/source/release_notes.rst:11]: nitpick: `primitive` -> `primitives`

File: featuretools/primitives/standard/transform/time_series/yeo_johnson.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        ColumnSchema(semantic_tags={"numeric"}),
+    ]
+    return_type = ColumnSchema(logical_type=Double, semantic_tags={"numeric"})
+    uses_full_dataframe = True
+
+    def __init__(self, lmbda=None):
+        self.lmbda = lmbda
+
+    def get_function(self):
+        def box_cox(numeric):
```
Comment (thehomebrewnerd) [inline featuretools/primitives/standard/transform/time_series/yeo_johnson.py:45]: This function should be renamed

File: featuretools/primitives/standard/transform/time_series/box_cox.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    uses_full_dataframe = True
+
+    def __init__(self, lmbda=None):
+        self.lmbda = lmbda
+
+    def get_function(self):
+        def box_cox(numeric):
+            if self.lmbda is not None:
+                return stats.boxcox(numeric, lmbda=self.lmbda)
+            return stats.boxcox(numeric)[0]
```
Comment (thehomebrewnerd) [inline featuretools/primitives/standard/transform/time_series/box_cox.py:49]: I'm not exactly sure how best to go about this, but how does this work in a train/test split scenario where a user might want to use this primitive on the training set and then separately later on a test set? I don't think there is any way for them to access the lambda value that was automatically determined so they could serialize the primitive and apply it to new data and have the new data transformed in the same way.

Also, if this primitive is applied to multiple numeric columns each column could end up with a unique lambda value, which presents additional challenges when knowing what lambda value actually got applied to each column.

I'm on the fence about whether we should automatically determine lambda or whether we should require the user to provide that value for now until we can sort this out better.

File: featuretools/primitives/standard/transform/time_series/box_cox.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class BoxCox(TransformPrimitive):
+    """Applies a transformation to help your data resemble a normal distribution,
+        commonly known as the Box Cox transformation. The input data must be positive.
+        It cannot contain zeroes or negative numbers.
+        Commonly used with time series data, to reduce the variance.
+
+    Args:
+        lmbda (int, float, optional): The exponent to use for the transformation.
```
Comment (thehomebrewnerd) [inline featuretools/primitives/standard/transform/time_series/box_cox.py:15]: If we leave this as an optional parameter, I think we should describe what happens in the case where it is not specified.

File: featuretools/primitives/standard/transform/time_series/yeo_johnson.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    name = "yeo_johnson"
+    input_types = [
+        ColumnSchema(semantic_tags={"numeric"}),
+    ]
+    return_type = ColumnSchema(logical_type=Double, semantic_tags={"numeric"})
+    uses_full_dataframe = True
+
+    def __init__(self, lmbda=None):
+        self.lmbda = lmbda
```
Comment (thehomebrewnerd) [inline featuretools/primitives/standard/transform/time_series/yeo_johnson.py:42]: Comments from `BoxCox` about lambda also apply here.

File: featuretools/primitives/standard/transform/time_series/box_cox.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+from woodwork.column_schema import ColumnSchema
+from woodwork.logical_types import Double
+
+from featuretools.primitives.base import TransformPrimitive
+
+
+class BoxCox(TransformPrimitive):
+    """Applies a transformation to help your data resemble a normal distribution,
+        commonly known as the Box Cox transformation. The input data must be positive.
+        It cannot contain zeroes or negative numbers.
```
Comment (thehomebrewnerd) [inline featuretools/primitives/standard/transform/time_series/box_cox.py:11]: Can't we offset the data if it has zero or negative values, prior to transforming? Although doing so, would present some of the same challenges that we have with the lambda values.

File: featuretools/tests/primitive_tests/test_transform_primitive.py [hunk 13]
```diff
@@ -711,10 +713,14 @@ def test_email_address_to_domain_empty_string():
 
 def test_email_address_to_domain_empty_series():
     email_address_to_domain = EmailAddressToDomain()
-    array = pd.Series([])
+    array = pd.Series([], dtype="category")
```
Comment (thehomebrewnerd) [inline featuretools/tests/primitive_tests/test_transform_primitive.py:716]: Not sure exactly why these changes are included, but the input to this primitive function will be an `EmailAddress` and shouldn't a column with `EmailAddress` logical type have a dtype of `string`?

File: featuretools/primitives/standard/transform/email/is_free_email_domain.py [hunk 14]
```diff
@@ -36,7 +36,7 @@ def get_function(self):
         def is_free_email_domain(emails):
             # if the input is empty return an empty Series
             if len(emails) == 0:
-                return pd.Series([])
+                return pd.Series([], dtype="boolean")
```
Comment (thehomebrewnerd) [inline featuretools/primitives/standard/transform/email/is_free_email_domain.py:39]: Why are these changes included?

If this is fixing something, maybe we should also include a note in the release notes?
